Tests/t_ldapobject.py

```diff
@@ -643,6 +646,66 @@ def test105_reconnect_restore(self):
             self.server._start_slapd()
         self.assertEqual(l1.whoami_s(), 'dn:'+bind_dn)
 
+    def test_106_reconnect_restore(self):
+        lo = self.ldap_object_class(self.server.ldap_uri, retry_max=2, retry_delay=1)
+        bind_dn = 'cn=user1,' + self.server.suffix
+        lo.simple_bind_s(bind_dn, 'user1_pw')
+
+        dn = lo.whoami_s()[3:]
+
+        self.server._proc.terminate()
+        self.server.wait()
+
+        # do a search, wait for the timeout, handle SERVER_DOWN exception
+        with self.assertRaises(ldap.SERVER_DOWN):
+            lo.search_s(dn, ldap.SCOPE_BASE, '(objectClass=*)')
+
+        self.server._start_slapd()
+
+        # try to use the connection again
+        lo.search_s(dn, ldap.SCOPE_BASE, '(objectClass=*)')
+        # this used to raise ldap.INSUFFICIENT_ACCESS, as it is became a unbound connection and did everything anonymously.
+
+    def test_107_reconnect_restore(self):
+        """The idea of this test is to restart the LDAP-Server while there are ongoing searches.
+        This causes :class:`ldap.UNAVAILABLE` to be raised (with |OpenLDAP|) for a short time.
+        To increase the chance of triggering this bug we are starting multiple threads
+        with a large number of retry attempts in a short amount of time.
+        """
+        # TODO: refactor this test
+        excs = []
+
+        def _reconnect_search_thread():
+            lo = self.ldap_object_class(self.server.ldap_uri)
+            bind_dn = 'cn=user1,' + self.server.suffix
+            lo.simple_bind_s(bind_dn, 'user1_pw')
+            lo._retry_max = 10E4
+            lo._retry_delay = .001
+            lo.search_ext_s(self.server.suffix, ldap.SCOPE_SUBTREE, "cn=user1", attrlist=["cn"])
+            s = time.time()
+            while (time.time() - s) < run_time:
+                lo.search_ext_s(self.server.suffix, ldap.SCOPE_SUBTREE, filterstr="cn=user1", attrlist=["cn"])
+
+        def reconnect_search_thread():
+            try:
+                _reconnect_search_thread()
+            except Exception as exc:  # ldap.UNAVAILABLE
```

Comment:
If `UNAVAILABLE` isn't being raised anymore, this comment is no longer valid.

Reply:
it was meant to tell what would be expected if this tests case was broken. i can remove it, if you want

Reply:
If you're not expecting a specific Exception as part of a successful test, then raising one on failure fails the test, correct? No need to catch and retrigger just for the sake of it.